### DISCUSSION TITLE
Loginpage translations override in Admin-UI message bundle

### DISCUSSION THREAD

**Discussion by seybsen** [keycloakify/keycloakify#401] [Q&A]
Hi :slightly_smiling_face: 

I'd like to override some of keycloak's default translations independently for diffrent realms.

I added two examples for german here:
![image](https://github.com/keycloakify/keycloakify/assets/739688/3df00492-e1d4-458e-80ed-4226f351797a)

... which are correctly shown in the keycloak theme:

![image](https://github.com/keycloakify/keycloakify/assets/739688/9618fd9d-50df-4fc7-b2ec-d1aadaa99aba)

... unfortunately those values are not used when switching to the keycloakify theme:

![image](https://github.com/keycloakify/keycloakify/assets/739688/d783cc6a-6216-46f9-8068-546f3d0d1141)

Any idea what could cause this?

**Comment by garronej** [comment]
Hello @seybsen,  

Sorry Keycloakfy does not take this into account. I wasn't aware of this feature.  

Can you do a right click on your page -> show source and paste the result here. I'll see if I can do something.

**Comment by garronej** [comment]
I anticipate that implementing this feature will not be a straightforward task. While it is achievable—given that we have successfully implemented similar functionalities—it will necessitate an additional build step.
I have already earmarked this feature in my roadmap, as I am committed to ensuring that the themes generated by Keycloakify remain uncompromising in their quality. My goal is for everything that functions seamlessly with a regular theme to work just as effectively with themes generated by Keycloakify.

That being said, for this specific use case, there may be alternative features within Keycloakify that would better suit your needs than overwriting the i18n messages from the Keycloak configuration.

Could you please provide more details about what exactly you are trying to accomplish? With a clearer understanding of your objectives, I may be able to suggest a more fitting approach for your situation.

**Comment by seybsen** [comment]
Hi @garronej,

thanks for your quick reply. We have a multi-tenant app which uses a keycloak realm for each customer. The customers have control over all translations from within the app itself and should be able to override translations from the keycloak login theme in that same UI too. We currently have no API to get translations from external services so we thought we could just use Keycloak's REST-API endpoint to manipulate those message bundles I mentioned in my initial question.

> Texts of these message bundles can be overwritten by realm-specific values. The realm-specific values are manageable via UI and API.

(https://www.keycloak.org/docs/latest/server_development/)

... unfortunately I can't find the endpoints in the docs at the moment (https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html)
I will take a closer look tomorrow (just on my phone right now).

Thanks for your work on this great project. We'd really like to get this awesome developer experience including a modern framework, storybook, and theme-variants and such 💪🏻😊

**Comment by garronej** [comment]
> thanks for your quick reply. We have a multi-tenant app which uses a keycloak realm for each customer. The customers have control over all translations from within the app itself and should be able to override translations from the keycloak login theme in that same UI too. We currently have no API to get translations from external services so we thought we could just use Keycloak's REST-API endpoint to manipulate those message bundles I mentioned in my initial question.  

That's a well-thought approach. At present, Keycloakify only supports customer customization through custom environment variables, similar to what is implemented for the [French Government Keycloak Theme](https://github.com/codegouvfr/keycloak-theme-dsfr#environment-variables). One workaround could be to instruct your users to specify the URL to a JSON file containing their translations.
I agree that this might not provide the best Developer Experience (DX), though.

> ... unfortunately I can't find the endpoints in the docs at the moment (https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html)
I will take a closer look tomorrow (just on my phone right now).  

If a suitable API endpoint is available, it would significantly simplify our approach.  


> Thanks for your work on this great project. We'd really like to get this awesome developer experience including a modern framework, storybook, and theme-variants and such 💪🏻😊  

You’re welcome, and thank you for the positive feedback!
I must admit, the theme-variant feature isn’t documented as thoroughly as it should be. To help bridge this gap, here’s a code snippet that enables loading different CSS main files based on the theme variant:

```ts

import { useMemo } from "react";

export function useThemeCss(themeName: string){
    useMemo(() => {
        switch(themeName){
            case "keycloakify-starter":
                // @ts-expect-error
                import("./keycloakify-starter.css");
                break;
            case "keycloakify-starter-variant-1":
                // @ts-expect-error
                import("./keycloakify-starter-variant-1.css");
                break;
        }
    }, [themeName]);
}
```
I’m planning to include this functionality in the starter, but unfortunately, time constraints have been a limiting factor.

Additionally, there is an issue regarding the increasing size of the .jar file when using theme variants. Thankfully, it doesn't affect runtime performance, and there is a straightforward solution. Again, it's on my list of things to address when time allows.

Please let me know if this is aligned with what you’re envisioning, and if there’s more I can do to assist you beside supprting the feature, wich I won't probably be able to do by the end of summer.
